dbms/src/IO/ReadHelpers.h

```diff
@@ -621,16 +630,51 @@ inline ReturnType readDateTimeTextImpl(time_t & datetime, ReadBuffer & buf, cons
         return readDateTimeTextFallback<ReturnType>(datetime, buf, date_lut);
 }
 
+template <typename ReturnType>
+inline ReturnType readDateTimeTextImpl(DateTime64 & datetime64, UInt32 scale, ReadBuffer & buf, const DateLUTImpl & date_lut)
+{
+    DB::DecimalComponents<DateTime64::NativeType> c;
+    readDateTimeTextImpl<void>(c.whole, buf, date_lut);
+
+    char separator;
+    if (buf.read(separator))
+    {
+        if (separator != '.')
+        {
+            return ReturnType(false);
+        }
+    }
+
+    auto remainder1 = buf.available();
+    readIntText(c.fractional, buf);
+    auto fractional_length = remainder1 - buf.available();
+
+    c.fractional *= common::exp10_i64(scale - fractional_length);
```

Comment:
Maybe `max(scale - fractional_length, 0)`?

Reply:
There was a bug, now it is fixed and DateTime64(3) can properly parse both '.123456' and '.12' fractional parts.